packages/backend/src/handlers/actions/quests/createCompletion/createCompletion.ts

```diff
@@ -0,0 +1,67 @@
+import { CreateQuestCompletionInput, CreateQuestCompletionOutput,
+  Quest_Completion_Insert_Input,
+  QuestRepetition_Enum,
+  QuestStatus_Enum, } from '../../../../lib/autogen/hasura-sdk';
+import { client } from '../../../../lib/hasuraClient';
+
+export async function createCompletion(
+  playerId: string,
+  questCompletion: CreateQuestCompletionInput,
+): Promise<CreateQuestCompletionOutput> {
+  if (!questCompletion.submission_link && !questCompletion.submission_text) {
+    throw new Error('Must provide at least a submission link or text');
+  }
+
+  const { quest_by_pk: quest } = await client.GetQuestById({ quest_id: questCompletion.quest_id });
+  if (!quest) {
+    throw new Error('Quest not found');
+  }
+
+  if (quest.status !== QuestStatus_Enum.Open) {
+    throw new Error('Quest must be open');
+  }
+
+  // Personal or unique, check if not already done by player
+  if (quest.repetition === QuestRepetition_Enum.Unique || quest.repetition === QuestRepetition_Enum.Personal) {
+    const { quest_completion: existingQuestCompletions } = await client.GetQuestCompletions({
+      player_id: playerId,
+      quest_id: questCompletion.quest_id,
+    });
+    if(existingQuestCompletions.length > 0) {
+      throw new Error('Player already completed this personal/unique quest');
+    }
+  }
+
+  // Recurring, check if not already done by player within cooldown
+  if (quest.repetition === QuestRepetition_Enum.Recurring && quest.cooldown) {
+    const { quest_completion: existingQuestCompletions } = await client.GetLastQuestCompletionForPlayer({
+      player_id: playerId,
+      quest_id: quest.id,
+    });
+    if(existingQuestCompletions.length > 0) {
+      const existingQuestCompletion = existingQuestCompletions[0];
+      const submittedAt = new Date(existingQuestCompletion.submitted_at);
+      const now = new Date();
+      const diff = +now - +submittedAt;
+      if(diff < (quest.cooldown*1000)) {
+        throw new Error('Player have to wait before being able to do this quest again');
```

Comment:
Should include the time diff:
```suggestion
        throw new Error(`You must wait ${diff / 1000} min before you can complete this quest again.`);
```

Ofcourse you'll want to format the time properly, I think we might have some utils for that otherwise moment.js can do it cleanly (e.g. automatically saying min or hrs depending on the length of time)

Reply:
That'd be great, but I don't like puting the time in the string, better in a variable. That'd imply a complex error management system that I'm too lazy to do in this PR/review. 
This error is just for safegarding purposes, the frontend shouldn't allow to do that, instead show a proper cooldown on the button with a dynamic timer to tell when the player can complete the quest again.